### PR TITLE
Implement 7z support, fix windows specific symlink issue and UI tests

### LINKED ISSUE #164: Support `.7z`
The title sums up the first part of the request pretty well, 7z is a pretty popular archive format, but 7z supports several different compression formats it supports, so I think it would also be good to add an option like `--format`/`-f` to specify a compression format independently of the file name, which would be useful in general too

### PR DESCRIPTION
**Pull request by Flat**
Full implementation of #393
Closes #164

### CODE REVIEW

**Review by marcospb19 — CHANGES_REQUESTED** [review]

**Review by marcospb19 — APPROVED** [review]
I applied the asked changes myself and made a small refactor. :+1: 

Thanks for your help, I really appreciate it!

**Comment by marcospb19** [comment]
Thanks for the help!

File: tests/ui.rs [hunk 1]
```diff
@@ -62,17 +62,29 @@ fn ui_test_err_compress_missing_extension() {
     let (_dropper, dir) = testdir().unwrap();
 
     // prepare
+    #[cfg(not(windows))]
     run_in(dir, "touch", "input").unwrap();
 
+    #[cfg(windows)]
+    run_in(dir, "cmd", "/C copy nul input").unwrap();
+
```
[the remark this answers is not in the document]
Comment (Flat) [reply tests/ui.rs]: This still runs the tests with `cargo test` and fails btw.

File: src/main.rs [hunk 2]
```diff
@@ -22,7 +22,7 @@ use utils::{QuestionAction, QuestionPolicy};
 const BUFFER_CAPACITY: usize = 1024 * 32;
 
 /// Current directory or empty directory
-static CURRENT_DIRECTORY: Lazy<PathBuf> = Lazy::new(|| env::current_dir().unwrap_or_default());
+static CURRENT_DIRECTORY: Lazy<PathBuf> = Lazy::new(|| std::fs::canonicalize(env::current_dir().unwrap_or_default()).unwrap_or_default());
```
[the remark this answers is not in the document]
Comment (Flat) [reply src/main.rs]: It doesn't. for example I have C:\Users\Username\Documents symlinked to I:\Documents, this turns into \\?i\Documents for file paths used, but current_dir returns C:\users\username\documents

**Comment by Flat** [comment]
Thanks for handling those changes and merging quickly! I think the stuff you reverted should be tested on windows though, while the testing isn't a big deal the symlink handling could be. If you need assistance testing that or somewhere else just let me know!

```
ouch on  main is 📦 v0.4.2 via 🦀 v1.76.0-nightly
❯ cargo test
   Compiling ouch v0.4.2 (C:\Users\Flat\Documents\dev\ouch)
    Finished test [unoptimized + debuginfo] target(s) in 1m 16s
     Running unittests src\main.rs (target\debug\deps\ouch-cccf543838ac19f3.exe)

running 5 tests
test extension::tests::builds_suggestion_correctly ... ok
test utils::formatting::tests::test_pretty_bytes_formatting ... ok
test extension::tests::test_extensions_from_path ... ok
test cli::args::tests::test_clap_cli_err ... ok
test cli::args::tests::test_clap_cli_ok ... ok

test result: ok. 5 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests\integration.rs (target\debug\deps\integration-837358a025335869.exe)

running 6 tests
test utils::src_is_not_tests - should panic ... ok
test utils::src_is_src ... ok
test unpack_rar ... ok
test multiple_files ... ok
test single_empty_file ... ok
test single_file ... ok

test result: ok. 6 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 10.45s

     Running tests\mime.rs (target\debug\deps\mime-3c31897f2e17fcf0.exe)

running 3 tests
test utils::src_is_not_tests - should panic ... ok
test utils::src_is_src ... ok
test sanity_check_through_mime ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.35s

     Running tests\ui.rs (target\debug\deps\ui-5370a5704da03aaf.exe)

running 8 tests
test utils::src_is_not_tests - should panic ... ok
test ui_test_err_decompress_missing_extension ... FAILED
test ui_test_ok_decompress ... FAILED
test ui_test_err_compress_missing_extension ... FAILED
test ui_test_ok_compress ... FAILED
test utils::src_is_src ... ok
test ui_test_usage_help_flag ... ok
test ui_test_err_missing_files ... ok

failures:

---- ui_test_err_decompress_missing_extension stdout ----
thread 'ui_test_err_decompress_missing_extension' panicked at tests\ui.rs:74:41:
called `Result::unwrap()` on an `Err` value: Error { kind: NotFound, message: "program not found" }

---- ui_test_ok_decompress stdout ----
thread 'ui_test_ok_decompress' panicked at tests\ui.rs:106:35:
called `Result::unwrap()` on an `Err` value: Error { kind: NotFound, message: "program not found" }

---- ui_test_err_compress_missing_extension stdout ----
thread 'ui_test_err_compress_missing_extension' panicked at tests\ui.rs:65:35:
called `Result::unwrap()` on an `Err` value: Error { kind: NotFound, message: "program not found" }

---- ui_test_ok_compress stdout ----
thread 'ui_test_ok_compress' panicked at tests\ui.rs:95:35:
called `Result::unwrap()` on an `Err` value: Error { kind: NotFound, message: "program not found" }


failures:
    ui_test_err_compress_missing_extension
    ui_test_err_decompress_missing_extension
    ui_test_ok_compress
    ui_test_ok_decompress

test result: FAILED. 4 passed; 4 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.04s

error: test failed, to rerun pass `--test ui`
```

**Comment by marcospb19** [comment]
@Flat 

> Thanks for handling those changes and merging quickly! I think the stuff you reverted should be tested on windows though, while the testing isn't a big deal the symlink handling could be. If you need assistance testing that or somewhere else just let me know!

That's curious... it passes in our Windows CI, but fails locally for you. I didn't know that. (Perhaps our CI uses WSL?)

---

I agree with you that it would be great to have UI testing for Windows, the problem is, I don't have a Windows machine and our CI isn't covering that case so that I can see the error when it happens.

Even if I could, the feedback loop of developing with CI can be painful, so I'm not sure if I'd be willing to do that, because I already struggle to give Ouch the attention it needs, so I'm reluctant to assume this responsibility.

Windows support right now looks like "tier 2 support", we (hopefully) know it compiles, and run most tests, but not all.

With that said, I'm willing to accept issues and PRs to solve that in case you want to do them, I just won't be able to help as much as I wish I could :v .

**Pull request merged**